### DISCUSSION TITLE
Using a medical camera for tracking with PLUS

### DISCUSSION THREAD

**Discussion by GbekeAdesiyun** [PlusToolkit/PlusLib#991] [Q&A] [answered]
Hi,

I'm working on a project that uses a medical camera for tracking with PLUS. I have an Epiphan frame grabber that I will connect to PLUS and the camera. Could you advise me on the right way to connect the camera to PLUS, as well as any way I can use PLUS for the tracking? 

Thanks,
Mogbekeloluwa Adesiyun

**Comment by Sunderlandkyl** [comment, marked as answer]
The Epiphan framegrabber should be supported by all Plus packages, which come with example config files that are ready to connect.

You can edit your config file to connect to a tracking device as well. Do you know what kind of tracker you will use?

Here is an example config file that connects to both a video and tracker device:
https://github.com/PlusToolkit/PlusLibData/blob/master/ConfigFiles/PlusDeviceSet_Server_Ultrasonix_C5-2_Ascension3DG_calibrated.xml

**Reply by GbekeAdesiyun** [reply]
Thank you. I plan on using computer vision techniques for tracking with a medical camera, so I'm not using a tracker. Do you know how I can achieve this with PLUS?

**Reply by Sunderlandkyl** [reply]
You can try using the Optical Marker Tracker to track the position of ArUco markers:
http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceOpticalMarkerTracker.html

**Reply by GbekeAdesiyun** [reply]
Thanks. What file can I edit if I choose to extend the functionality of the Optical Marker Tracker?

**Reply by Sunderlandkyl** [reply]
If you plan on adding a new type of computer vision tracking to Plus, then you should add a new device.
You can find a document on the steps required here: https://plustoolkit.github.io/devicecode.

**Reply by GbekeAdesiyun** [reply]
Seems like the code to add the device has to be in C++, can this be written in Python instead?

**Reply by Sunderlandkyl** [reply]
All code in Plus is written in C++. If you don't want to try writing your device in C++, you can send the images from Plus to python using [PyIGTL](https://github.com/lassoan/pyigtl). PyPI link: https://pypi.org/project/pyigtl/.

**Reply by GbekeAdesiyun** [reply]
It seems like there is an issue with the optical tracking system on Slicer. My webcam shows on Slicer, but the arUco markers aren't being detected. Do you know what might be the cause of this issue? I have attached the screenshot of my I/O configuration on Slicer and the Plus xml config file. 
<img width="502" alt="image" src="https://user-images.githubusercontent.com/66129702/195444159-534f7931-3adb-4177-bec6-4558dd554c3d.png">

<img width="910" alt="image" src="https://user-images.githubusercontent.com/66129702/195445161-64e66b8e-e524-4c48-914e-62f5fdd92306.png">
<img width="886" alt="image" src="https://user-images.githubusercontent.com/66129702/195445214-95edd0b9-ccd7-429b-bbb4-aaf731aa3c7e.png">

<img width="882" alt="image" src="https://user-images.githubusercontent.com/66129702/195445312-03a081eb-1244-4282-a564-af9e552ab705.png">

**Reply by Sunderlandkyl** [reply]
Some webcams flip the orientation of the image, which can break the marker detection.
Try changing the PortUsImageOrientation to UN. Valid orientations are UN, UF, MN, MF.

**Reply by GbekeAdesiyun** [reply]
Thanks. I already changed the orientation to MN as shown in the screenshot. Also, the image captured by my webcam isn't flipped; it's oriented correctly, so the only issue I'm having is with the arUco markers not being detected. Any ideas on what I can do to fix this?

**Reply by GbekeAdesiyun** [reply]
Hi, I'm just following up regarding this

**Reply by Sunderlandkyl** [reply]
If you connect to the Image stream in Slicer from 18945, your image should appear upside down in the red slice view, and you should be able to read text if you hold it upside down. The image orientation that works for me is MF. Have you tried all possible orientations?

If you've tried the above, could you attach the log file?

**Reply by GbekeAdesiyun** [reply]
Thanks. The optical tracker actually works with the current orientation for my webcam on Plus. However, the optical tracker still doesn't work with my medical camera, and my image doesn't even show up even after I changed the image orientation. I have attached the log file 
[log_file.txt](https://github.com/PlusToolkit/PlusLib/files/9789139/log_file.txt)

**Reply by GbekeAdesiyun** [reply]
Hi, do you have any ideas on what I can do to fix this? This is the error message I am getting now
![image](https://user-images.githubusercontent.com/66129702/196774517-135dd299-6699-4333-bdb3-389c836917e5.png)

**Reply by Sunderlandkyl** [reply]
> However, the optical tracker still doesn't work with my medical camera, and my image doesn't even show up even after I changed the image orientation. 

Nothing looks wrong in the log file does the volume node show up in the data module in Slicer?

> This is the error message I am getting now

Can you attach a log file with the error?

**Reply by GbekeAdesiyun** [reply]
Could you clarify what you mean by volume node? I have attached the log file
[logFile19102022.txt](https://github.com/PlusToolkit/PlusLib/files/9823832/logFile19102022.txt)

**Reply by Sunderlandkyl** [reply]
> Could you clarify what you mean by volume node?

It should look like this in the Slicer Data module:
![image](https://user-images.githubusercontent.com/9222709/197001667-26122a24-8587-4d71-855f-715c8f291c5a.png)

I don't see the cause of the error in the log. Could you attach the config file causing the error as a txt?

**Reply by GbekeAdesiyun** [reply]
In that case, the volume node does show up. For sure, this is the config file. 
[configFile20102022.txt](https://github.com/PlusToolkit/PlusLib/files/9833674/configFile20102022.txt)

**Reply by Sunderlandkyl** [reply]
Do you have connections open in Slicer for both port 18944 (tracking) and 18945 (video)?

**Reply by Sunderlandkyl** [reply]
I'm not sure why the server is crashing. Does it still work if you use a config file with only the webcam device?

**Reply by GbekeAdesiyun** [reply]
Yes, I have connections for both of those ports. The IGTL type for both of those ports is "WAIT" though. Yes, it works with my webcam.

**Reply by Sunderlandkyl** [reply]
Ok, can you set your log level to trace, and then try to connect again.

After, you should find 2 log files that were modified in "C:\Users\gbeke\PlusApp-2.8.0.20191105-Win64\data", one for the launcher, and one for the server. Can you send me both of them?

**Reply by GbekeAdesiyun** [reply]
Okay, I just did that. These are the two log files: 
[102022_164608_PlusLog.txt](https://github.com/PlusToolkit/PlusLib/files/9833796/102022_164608_PlusLog.txt)
[102022_164613_PlusLog.txt](https://github.com/PlusToolkit/PlusLib/files/9833797/102022_164613_PlusLog.txt)

**Reply by Sunderlandkyl** [reply]
The camera listed is: "DVI2USB 3.0 D2S312174" and looks like it isn't compatible with the 640 480 frame size.
These are the configurations that are listed:
> Frame size: 1920x1080, video format: RGB24, frame rate: 15625
> Frame size: 1920x1080, video format: UYVY, frame rate: 15625
> Frame size: 1920x1080, video format: YUY2, frame rate: 15625
> Frame size: 1920x1080, video format: I420, frame rate: 312500
> Frame size: 1920x1080, video format: YV12, frame rate: 312500
> Frame size: 1920x1080, video format: NV12, frame rate: 312500
> Frame size: 1920x1080, video format: RGB565, frame rate: 15625
> Frame size: 1920x1080, video format: RGB8, frame rate: 312500

Try setting your FrameSize to 1920x1090.

**Reply by GbekeAdesiyun** [reply]
The camera feed is now being displayed on Slicer; thank you ! I'll test the optical tracker with the arUco markers now

**Reply by GbekeAdesiyun** [reply]
I can see the camera feed in slicer but it seems like the optical tracker still isn't detecting the arUco markers. What can I do to resolve this issue?

**Reply by Sunderlandkyl** [reply]
If your camera orientation is correct (image should appear upside down in the red slice view, and you should be able to read text if you hold it upside down), then it could be a problem with your markers.

Could you attach a picture of your physical marker and camera setup?
Could you also attach the PlusServer log file from "...\PlusApp-2.8.0.20191105-Win64\data"?

**Reply by GbekeAdesiyun** [reply]
I see. I have attached the log file and the pictures.
[PlusServerLogFile27102022.txt](https://github.com/PlusToolkit/PlusLib/files/9880750/PlusServerLogFile27102022.txt)

![image](https://user-images.githubusercontent.com/66129702/198345880-0f12cccc-2643-4813-a296-75ffc770dd6d.png)
![image](https://user-images.githubusercontent.com/66129702/198345951-24b592ca-1269-4ef4-b5f8-94e329c4446c.png)
![image](https://user-images.githubusercontent.com/66129702/198346113-dd24f467-61f0-4255-bcf0-79b3b031d310.png)

**Reply by Sunderlandkyl** [reply]
Ok, lets try the aruco_test application to see if the markers are detected at all.

If you open command prompt and navigate to "...\PlusApp-2.8.0.20191105-Win64\bin" and run aruco_test.exe:
> aruco_test.exe live:0 -d ARUCO_MIP_36h12

It should look like this if the marker is detected:
![image](https://user-images.githubusercontent.com/9222709/198354170-edb760f1-29c2-4225-821a-feaafe80f83a.png)

**Reply by GbekeAdesiyun** [reply]
I didn't get that image, I just got the result of basic image thresholding with no bounding boxes after running that executable. Could the issue be the fact that my markers are too big?

**Reply by Sunderlandkyl** [reply]
There should be two windows when you run the executable, one that exposes the thresholding parameters/webcam video/detected markers, and one that shows the thresholded images.

They may not be detected if they are too close/too far from the camera.

**Reply by GbekeAdesiyun** [reply]
I was referring to the thresholded image. I can also see the window that exposes the thresholding parameters. Are you saying the size doesn't affect the detection capability and the issue only pertains to the distance between the marker and the camera?

**Reply by Sunderlandkyl** [reply]
My experience has been that larger markers need to be farther away from the camera.

Unfortunately I am not that familiar with the factors that affect marker detection, so I am not going to be of much help for improving it.

**Reply by GbekeAdesiyun** [reply]
I see. That's alright

**Reply by GbekeAdesiyun** [reply]
I tried moving the markers farther away from the camera but that didn't fix the issue. I've also tried using smaller markers, but that also didn't work. Is there anything else I can do to resolve this?

**Reply by GbekeAdesiyun** [reply]
Hi, I'm just following up to find out the next steps to take

**Reply by Sunderlandkyl** [reply]
When you printed the markers, did you follow the instructions listed here (print in actual size: scaling should be 100%): http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceOpticalMarkerTracker.html?

**Reply by GbekeAdesiyun** [reply]
I see; I was actually using a printout that did not have all the markers scaled to 100%. I had already printed out the aruco markers from the pdf given in the link you sent, but I had not tested that yet. The aruco markers are now being detected now that I have changed the printout. Thanks.
